theia-extensions/viewer-prototype/src/browser/trace-server-url-provider-frontend-impl.ts

```diff
@@ -1,66 +1,126 @@
+        @inject(MessageService) protected messageService: MessageService,
+    ) {
+        this._traceServerUrlPromise = new Promise(resolve => {
+            const self = this.onDidChangeTraceServerUrl(url => {
+                self.dispose();
+                resolve(url);
+            });
+        });
+        // Get the URL template from the remote environment.
+        this.environment.getValue('TRACE_SERVER_URL').then(variable => {
+            const url = variable?.value;
+            this._traceServerUrlTemplate = url
+                ? this.normalizeUrl(url)
+                : TRACE_SERVER_DEFAULT_URL;
+            this.updateTraceServerUrl();
+        });
+        // Get the configurable port from Theia's preferences.
+        this.tracePreferences.ready.then(() => {
+            this._traceServerPort = this.tracePreferences[TRACE_PORT];
+            this.updateTraceServerUrl();
+            this.tracePreferences.onPreferenceChanged(async event => {
+                if (event.preferenceName === TRACE_PORT) {
+                    this._traceServerPort = event.newValue;
+                    this.updateTraceServerUrl();
+                    try {
+                        await this.traceServerConfigService.stopTraceServer();
+                        this.messageService.info(`Trace server disconnected on port: ${event.oldValue}.`);
+                    } catch (_) {
+                        // Do not show the error incase the user tries to modify the port before starting a server
+                    }
+                }
+            });
+        });
     }
 
-    async onStart(_app: FrontendApplication): Promise<void> {
-        this._traceServerUrl = await this.traceViewerEnvironment.getTraceServerUrl();
-        this.updateListeners();
+    async initialize(): Promise<void> {
+        // Don't start the application until the Trace Server URL is initialized.
+        await this._traceServerUrlPromise;
     }
 
-    async updateListeners(): Promise<void> {
-        this._listeners.forEach(listener => listener(this._traceServerUrl));
+    async getTraceServerUrlPromise(): Promise<string> {
+        return this._traceServerUrlPromise;
     }
 
     getTraceServerUrl(): string {
+        if (this._traceServerUrl === undefined) {
+            throw new Error('The Trace Server URL is not yet defined. Try using getTraceServerUrlPromise.');
+        }
         return this._traceServerUrl;
     }
 
-    /**
-     * Add a listener for trace server url changes
-     * @param listener The listener function to be called when the url is
-     * changed
-     */
-    addTraceServerUrlChangedListener(listener: (url: string) => void): void {
-        this._listeners.push(listener);
+    protected normalizeUrl(url: string): string {
+        url = url.toLowerCase();
+        // Add missing http protocol.
+        if (!url.endsWith('http://') || !url.endsWith('https://')) {
```

Comment:
This if statement needs to be like this:
if (!url.startsWith('http://') && !url.startsWith('https://'))

Reply:
I always get this wrong, good catch.

Reply:
Even with the original code it would have failed now because GitPod uses https now